content/en/blog/2022-08-05-effective-programmers-read-code.md

```diff
@@ -0,0 +1,59 @@
+
+The primary purpose of the glass castle is to ensure that we can quickly locate related code in the codebase and understand the consequences of making changes to the code. It does serve another purpose however: like [code "smells"](https://martinfowler.com/bliki/CodeSmell.html), I use the glass castle to understand when there is a structural or architectural problem with the system. For example, when I zoom into a component and I see that there are too many external interactions with subcomponents, it's an indication that the parent component is not correctly abstracted. This is very similar to the idea of ["wide vs deep" modules](https://nakabonne.dev/posts/depth-of-module/) described in Ousterhout's [_A Philosophy of Software Design_](https://web.stanford.edu/~ouster/cgi-bin/book.php). The best use of the glass castle is to determine when the system is "tipping" over as we add features or make changes because it may lead us to refactoring or addressing the initial design that wasn't quite right for the current iteration of software.
+
+The glass castle is essential for how I write good code and recently, I've started to pay more attention to how I construct the glass castle. When I sit down to implement a feature or make a change, the very first thing I do is identify candidate locations in the code where the change needs to be made. These locations are identified in a top down fashion looking at the organization of the code and quickly scanning through package and module names and package documentation.
+
+Once I find the candidate locations, I continue in a bottom up fashion identifying the functional dependencies surrounding the location and putting together components and interactions tagged with the input and output data contracts that they make. Often, I have to read the code of the external dependencies of the project to figure out the best possible solutions. I continue grouping, linking, and encapsulating code in my head until I feel like I have enough of a picture that I'm confident in the change I'm about to make.
+
+Before I've written even a single line of code, I've read hundreds of lines of code. The more complex the change or the system (or the longer I was away from the system before making a change), the more time I have to spend reading code before writing it. That's when I realized that as a programmer I spend far more time reading code than writing it.[^3]
```

Comment:
Suggestion: "Before I've written even a single line of code, I've already read hundreds of lines of code. The more complex the change or system (or, the longer I've been away from the system before making a change),..." 

"It's through situations like these that I've come to realize, as a programmer, I spend far more time reading code than writing it."